app/models/participant_profile/mentor.rb

```diff
@@ -22,9 +22,44 @@ class ParticipantProfile::Mentor < ParticipantProfile::ECF
   }
 
   def self.archivable_from_frozen_cohort(restrict_to_participant_ids: [])
-    super(restrict_to_participant_ids:)
-      .where(mentor_completion_date: nil)
-      .where.not(id: InductionRecord.where.not(mentor_profile_id: nil).select(:mentor_profile_id).distinct)
+    archivable_states = %i[ineligible voided submitted].freeze
+
+    # Mentors that have no FIP induction records
+    not_fip = InductionRecord.joins(:induction_programme, participant_profile: { schedule: :cohort })
+                             .where.not(cohorts: { payments_frozen_at: nil })
+                             .where(participant_profiles: { mentor_completion_date: nil })
+                             .where(participant_profiles: { type: "ParticipantProfile::Mentor" })
+                             .where.not(induction_programme: { training_programme: :full_induction_programme })
+    not_fip = not_fip.where(participant_profile_id: restrict_to_participant_ids) if restrict_to_participant_ids.any?
+    not_fip_ids = not_fip.pluck(:participant_profile_id).uniq
+
+    # Mentors that have at least one non-archivable declaration
+    with_unarchivable_declaration = joins(:participant_declarations, schedule: :cohort)
+                                      .where.not(cohorts: { payments_frozen_at: nil })
+                                      .where(mentor_completion_date: nil)
+                                      .where.not(id: not_fip_ids)
+                                      .where.not(participant_declarations: { state: archivable_states })
+                                      .distinct
+    with_unarchivable_declaration = with_unarchivable_declaration.where(id: restrict_to_participant_ids) if restrict_to_participant_ids.any?
```

Comment:
It looks like a lot of the logic in the two subclasses is very similar; can any of it be extracted to separate methods to DRY it up?

Reply:
I agree Ross but I can't expend more time on this. I should be working on ECF2 instead.

Reply:
I would have liked to just modify the method you wrote in the parent class but it all has performance issues (actually and exception was thrown in production when I tested it) and I had to include all sort of restrictions on each sub-query to be able to get reasonable response times.